File: DSA Practice/LeetCode/451. Sort Characters By Frequency/solution.java
class Solution {
    public String frequencySort(String s) {
        HashMap<Character, Integer> freq = new HashMap<Character, Integer>();

        for (char c : s.toCharArray()){
            if (freq.containsKey(c)){
                freq.put(c, freq.get(c) + 1);
            } else {
                freq.put(c, 1);
            }
        }

        StringBuilder sb = new StringBuilder();
        int m = s.length();

        while (sb.length() < s.length()){
            for (Map.Entry e : freq.entrySet()){
                if ((int) e.getValue() == m){
                    for (int i = 0; i < (int) e.getValue(); i++){
                        sb.append((char) e.getKey());
                    }

                    freq.put((char) e.getKey(), -1);
                }
            }

            m--;
        }

        return sb.toString();
    }
}

File: DSA Practice/LeetCode/451. Sort Characters By Frequency/solution.kt
class Solution {
    fun frequencySort(s: String): String {
        val freq = mutableMapOf<Char, Int>()

        for (c in s){
            if (freq.containsKey(c)){
                freq.put(c, freq.get(c)!!.plus(1))
            } else {
                freq.put(c, 1)
            }
        }

        var m = s.length
        var sb = StringBuilder()

        while(sb.length < s.length){
            for ((k, v) in freq){
                if (v == m){
                    for (i in 0..v-1){
                        sb.append(k)
                    }
                    freq.put(k, -1)
                }
            }

            m--
        }

        return sb.toString()
        
    }
}